crypto/src/openpgp/PgpSecretKey.cs

```diff
@@ -1252,51 +1288,68 @@ internal static PgpSecretKey DoParseSecretKeyFromSExpr(Stream inputStream, byte[
             throw new PgpException("unknown key type found");
         }
 
-        private static byte[] GetDValue(Stream inputStream, byte[] rawPassPhrase, bool clearPassPhrase, string curveName)
+        private static byte[] GetDValue(SXprUtilities reader, byte[] rawPassPhrase, bool clearPassPhrase, string curveName)
         {
             string type;
-            SXprUtilities.SkipOpenParenthesis(inputStream);
+            reader.SkipOpenParenthesis();
 
             string protection;
             S2k s2k;
             byte[] iv;
             byte[] secKeyData;
 
-            type = SXprUtilities.ReadString(inputStream, inputStream.ReadByte());
+            type = reader.ReadString();
             if (type.Equals("protected"))
             {
-                protection = SXprUtilities.ReadString(inputStream, inputStream.ReadByte());
+                protection = reader.ReadString();
 
-                SXprUtilities.SkipOpenParenthesis(inputStream);
+                reader.SkipOpenParenthesis();
 
-                s2k = SXprUtilities.ParseS2k(inputStream);
+                s2k = reader.ParseS2k();
 
-                iv = SXprUtilities.ReadBytes(inputStream, inputStream.ReadByte());
+                iv = reader.ReadBytes();
 
-                SXprUtilities.SkipCloseParenthesis(inputStream);
+                reader.SkipCloseParenthesis();
 
-                secKeyData = SXprUtilities.ReadBytes(inputStream, inputStream.ReadByte());
+                secKeyData = reader.ReadBytes();
             }
             else
             {
                 throw new PgpException("protected block not found");
             }
 
-            // TODO: recognise other algorithms
-            KeyParameter key = PgpUtilities.DoMakeKeyFromPassPhrase(SymmetricKeyAlgorithmTag.Aes128, s2k, rawPassPhrase, clearPassPhrase);
+            // Valid values of protection: openpgp-s2k3-sha1-aes-cbc, openpgp-s2k3-ocb-aes, openpgp-native
+            byte[] data;
+            KeyParameter key;
+
+            switch (protection)
+            {
+                case "openpgp-s2k3-sha1-aes-cbc":
+                    key = PgpUtilities.DoMakeKeyFromPassPhrase(SymmetricKeyAlgorithmTag.Aes128, s2k, rawPassPhrase, clearPassPhrase);
+                    data = RecoverKeyData(SymmetricKeyAlgorithmTag.Aes128, "/CBC/NoPadding", key, iv, secKeyData, 0, secKeyData.Length);
+                    break;
 
-            byte[] data = RecoverKeyData(SymmetricKeyAlgorithmTag.Aes128, "/CBC/NoPadding", key, iv, secKeyData, 0, secKeyData.Length);
+                case "openpgp-s2k3-ocb-aes":
+                    key = PgpUtilities.DoMakeKeyFromPassPhrase(SymmetricKeyAlgorithmTag.Aes128, s2k, rawPassPhrase, clearPassPhrase);
+                    data = RecoverKeyData(SymmetricKeyAlgorithmTag.Aes128, "/OCB/NoPadding", key, iv, secKeyData, 0, secKeyData.Length);
```

Comment:
This doesn't work properly since it fails to verify the MAC in the OCB transformation. The decrypted data are correct though.

Reply:
Looks like I missed passing the associated authentication data (https://github.com/gpg/gnupg/blob/f541e1d95a91d4764c7ed0df10c293bfd493dd41/agent/protect.c#L478-L484).